### PR TITLE
chore: install docker built extensions on AMI

### LINKED ISSUE #513: Build Postgres bundled in Docker image from source to achieve parity with the AMI
The official Postgres Docker image, off of which we base our Docker builds, is built using different compiler flags than the ones we're using to build Postgres in the AMI, which can create problems for users developing locally and then moving to projects created on the platform if they're using features enabled by default by the the Docker image, as they're missing in the AMI which is deployed on the platform.

### PR DESCRIPTION
**Pull request by sweatybridge**
## What kind of change does this PR introduce?

closes https://github.com/supabase/postgres/issues/513

## What is the current behavior?

Postgres AMI is built with a [subset of flags](https://github.com/supabase/postgres/issues/511) used in docker image

## What is the new behavior?

- Install AMI postgres from the official Postgres PPA
- Install AMI extensions from packages built in docker

## Additional context

TODO:
- [ ] delete unused playbooks

### CODE REVIEW

File: ansible/tasks/setup-postgres.yml
Comment (pcnc) [inline ansible/tasks/setup-postgres.yml:135]: Removing these folders would interfere with adminapi's [optimizations](https://github.com/supabase/supabase-admin-api/blob/master/optimizations/postgres.go), amongst other features

File: ansible/tasks/setup-postgres.yml
Comment (pcnc) [inline ansible/tasks/setup-postgres.yml:98]: Is the Postgres provided through the PPA compiled with the same cflags we were compiling Postgres with? - or at least additive wrt flags.

File: ansible/tasks/setup-postgres.yml
Comment (pcnc) [inline ansible/tasks/setup-postgres.yml:124]: Should we ensure the newly created postgres user is also part of the `ssl-cert` group?

**Review by pcnc — CHANGES_REQUESTED** [review]

File: ansible/tasks/setup-postgres.yml
Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:135]: These directories are created by the next step so should be a duplicate.

File: ansible/tasks/setup-postgres.yml
Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: Yup the official docker image downloads from PPA and are additive to our flags. 

I will check again later

File: ansible/tasks/setup-postgres.yml
Comment (pcnc) [reply ansible/tasks/setup-postgres.yml:135]: Ah, didn't see that

File: ansible/tasks/setup-postgres.yml
Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:124]: I just checked that the default apt installation adds postgres user to `ssl-cert` group.

```bash
$ docker run --rm -t supabase/postgres:15.1.0.84 groups postgres
postgres : postgres ssl-cert
```

File: ansible/tasks/setup-postgres.yml
Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: - Build configs from arm64 PPA:
```bash
$ docker run --rm -t supabase/postgres:15.1.0.84 pg_config
BINDIR = /usr/lib/postgresql/15/bin
DOCDIR = /usr/share/doc/postgresql-doc-15
HTMLDIR = /usr/share/doc/postgresql-doc-15
INCLUDEDIR = /usr/include/postgresql
PKGINCLUDEDIR = /usr/include/postgresql
INCLUDEDIR-SERVER = /usr/include/postgresql/15/server
LIBDIR = /usr/lib/aarch64-linux-gnu
PKGLIBDIR = /usr/lib/postgresql/15/lib
LOCALEDIR = /usr/share/locale
MANDIR = /usr/share/postgresql/15/man
SHAREDIR = /usr/share/postgresql/15
SYSCONFDIR = /etc/postgresql-common
PGXS = /usr/lib/postgresql/15/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE =  '--build=aarch64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/aarch64-linux-gnu' '--runstatedir=/run' '--disable-maintainer-mode' '--disable-dependency-tracking' '--with-tcl' '--with-perl' '--with-python' '--with-pam' '--with-openssl' '--with-libxml' '--with-libxslt' '--mandir=/usr/share/postgresql/15/man' '--docdir=/usr/share/doc/postgresql-doc-15' '--sysconfdir=/etc/postgresql-common' '--datarootdir=/usr/share/' '--datadir=/usr/share/postgresql/15' '--bindir=/usr/lib/postgresql/15/bin' '--libdir=/usr/lib/aarch64-linux-gnu/' '--libexecdir=/usr/lib/postgresql/' '--includedir=/usr/include/postgresql/' '--with-extra-version= (Ubuntu 15.3-1.pgdg20.04+1)' '--enable-nls' '--enable-thread-safety' '--enable-debug' '--enable-dtrace' '--disable-rpath' '--with-uuid=e2fs' '--with-gnu-ld' '--with-gssapi' '--with-ldap' '--with-pgport=5432' '--with-system-tzdata=/usr/share/zoneinfo' 'AWK=mawk' 'MKDIR_P=/bin/mkdir -p' 'PROVE=/usr/bin/prove' 'PYTHON=/usr/bin/python3' 'TAR=/bin/tar' 'XSLTPROC=xsltproc --nonet' 'CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now' '--enable-tap-tests' '--with-icu' '--with-llvm' 'LLVM_CONFIG=/usr/bin/llvm-config-10' 'CLANG=/usr/bin/clang-10' '--with-lz4' '--with-zstd' '--with-systemd' '--with-selinux' 'build_alias=aarch64-linux-gnu' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security'
CC = gcc
CPPFLAGS = -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -moutline-atomics -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -L/usr/lib/llvm-10/lib -Wl,--as-needed
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lselinux -lzstd -llz4 -lxslt -lxml2 -lpam -lssl -lcrypto -lgssapi_krb5 -lz -lreadline -lpthread -lrt -ldl -lm
```

- Build configs from AMI:
```bash
ubuntu@db-cimpzzikygouamvsjwxa:~$ pg_config
BINDIR = /usr/lib/postgresql/bin
DOCDIR = /var/lib/postgresql/doc
HTMLDIR = /var/lib/postgresql/doc
INCLUDEDIR = /usr/local/pgsql/include
PKGINCLUDEDIR = /usr/local/pgsql/include
INCLUDEDIR-SERVER = /usr/local/pgsql/include/server
LIBDIR = /usr/lib/postgresql/lib
PKGLIBDIR = /usr/lib/postgresql/lib
LOCALEDIR = /var/lib/postgresql/locale
MANDIR = /var/lib/postgresql/man
SHAREDIR = /var/lib/postgresql
SYSCONFDIR = /usr/local/pgsql/etc
PGXS = /usr/lib/postgresql/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE =  '--with-llvm' '--with-openssl' '--with-systemd' '--with-uuid=e2fs' '--with-libxml' '--with-icu' '--with-lz4' '--exec-prefix=/usr/lib/postgresql' '--datarootdir=/var/lib/postgresql' 'CFLAGS=-moutline-atomics -mtune=native -march=native -mcpu=native -fsigned-char -O2' 'LLVM_CONFIG=/usr/bin/llvm-config-11' 'CLANG=/usr/bin/clang-11'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -moutline-atomics -mtune=native -march=native -mcpu=native -fsigned-char -O2
CFLAGS_SL = -fPIC
LDFLAGS = -L/usr/lib/llvm-11/lib -Wl,--as-needed -Wl,-rpath,'/usr/lib/postgresql/lib',--enable-new-dtags
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -llz4 -lxml2 -lssl -lcrypto -lz -lreadline -lpthread -lrt -ldl -lm
```

The following CFLAGS are used by our AMI build but not by official PPA
1. `-mtune=native -march=native -mcpu=native`
2. `-fsigned-char`

The `mtune`, `march`, and `mcpu` flags optimise for CPU of the **builder** machine so we _might_ see a performance boost if the AMI is run on the same AWS instance type as the builder. Given the various compute sizes offered by AWS for upgrades, its benefit really varies from project to project. Hence, I'm in favour of dropping these flags for a more generic arm64 build.

The `signed-char` flag seems to be the [default for x86 but not for arm64](https://news.ycombinator.com/item?id=18269886). I don't think it affects the integrity of the compiled binary because the official package is built in a platform specific way. Tagging @dragarcia just in case I missed anything.

File: .github/workflows/test.yml
Comment (sweatybridge) [inline .github/workflows/test.yml]: delete after approve

File: .github/workflows/ami-release.yml
Comment (sweatybridge) [inline .github/workflows/ami-release.yml]: delete after approve

File: ansible/tasks/setup-postgres.yml
Comment (pcnc) [reply ansible/tasks/setup-postgres.yml:98]: We're using Graviton3 machines to build AMIs and run projects, which all run of Neoverse-N1 CPUs 🤔 would be curious to run a few benchmarks just to see if there's any noticeable difference.

Will test `15.1.0.85-rc0` on staging against the currently deployed image and circle back!

Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: Also interesting to note that the official debian package applies a series of patches on top of original source during build. For eg. the `-moutline-atomics` flag is enabled through [a patch](https://salsa.debian.org/postgresql/postgresql/-/blob/15/debian/patches/focal-arm64-outline-atomics). I'm not sure if things are this way due to legacy reasons but I want to have good faith in postgres PPA maintainer for improving the deb distribution. It also helps to lighten our maintenance burden.

So overall my preference is to stick with the flags used by official distribution (if performance difference is negligible). If we find anything wrong or missing with the PPA, we can consider asking in their mailing list and possibly contribute our changes upstream.

Comment (pcnc) [reply ansible/tasks/setup-postgres.yml:98]: The PPA-sourced Postgres seems to have higher throughput for `tcpb-like` but looses out on `select-only` and `simple-update` 🤔

tcpb-like
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 55968
number of failed transactions: 0 (0.000%)
latency average = 106.504 ms
initial connection time = 614.456 ms
tps = 938.928527 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 47349
number of failed transactions: 0 (0.000%)
latency average = 125.906 ms
initial connection time = 592.728 ms
tps = 794.244696 (without initial connection time)
```

select-only
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 909330
number of failed transactions: 0 (0.000%)
latency average = 6.538 ms
initial connection time = 635.066 ms
tps = 15294.214961 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 986558
number of failed transactions: 0 (0.000%)
latency average = 6.031 ms
initial connection time = 595.715 ms
tps = 16581.834796 (without initial connection time)
```

simple-update
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b simple-update
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 343008
number of failed transactions: 0 (0.000%)
latency average = 17.345 ms
initial connection time = 616.760 ms
tps = 5765.515337 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b simple-update
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 376560
number of failed transactions: 0 (0.000%)
latency average = 15.807 ms
initial connection time = 588.914 ms
tps = 6326.249293 (without initial connection time)
```

Differences are non-trivial in either case - might be that there are some improvements bundled with PG 15.3 which can help in a mixed workload the sorts the tcpb-like suite runs.
I'd be up to getting the postgres team onboard to reach out through the mailing list so we can contribute with a Graviton-specific patch!

Regarding this - it seems that by installing the PPA we're moving to Postgres 15.3, rather than keeping 15.1 (thus diverging from the `postgres_release` variable). 

I also see that we're pinning the version:
```
postgresql-{{ postgresql_major }}={{ postgresql_release }}-1.pgdg20.04+1
``` 

but for some reason this doesn't seem to be respected?

`apt list` confirms that 15.3 is installed. Instance ref is  `nfouvluloswtodcmjimp` on staging.
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ apt list --installed | grep postgresql
postgresql-15/now 15.3-1.pgdg20.04+1 arm64 [installed,local]
postgresql-client-15/now 15.3-1.pgdg20.04+1 arm64 [installed,local]
postgresql-client-common/now 249.pgdg20.04+1 all [installed,local]
postgresql-common/now 249.pgdg20.04+1 all [installed,local]
```

Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: > but for some reason this doesn't seem to be respected?

Hmm good catch. My local docker container is reporting 15.1 (for server which is what we care most about) so it could be due to some differences in ami step. I will look for a fix.

```bash
root@02e0ee680060:/# apt list --installed | grep postgresql

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

postgresql-15/now 15.1-1.pgdg20.04+1 amd64 [installed,local]
postgresql-client-15/now 15.3-1.pgdg20.04+1 amd64 [installed,local]
postgresql-client-common/now 249.pgdg20.04+1 all [installed,local]
postgresql-common/now 249.pgdg20.04+1 all [installed,local]
postgresql-server-dev-15/now 15.1-1.pgdg20.04+1 amd64 [installed,local]
```

EDIT: I read [ansible documentation](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/apt_module.html#examples) again and seems like the suffix `-1.pgdg20.04+1` is unnecessary. Will rebuild and circle back.

Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: I managed to obtain similar results on m6g.xlarge for pg 15.1, which is built from the latest commit here. I used AMI version `15.1.0.86-rc1` and ran a few times for consistency.

Average latency for tspb-like workloads is ~10% faster on pre-built than custom (112ms vs 125ms). Although simple select and update are ~10% slower for pre-built, I still think they are negligible because their absolute values are small, ie. 6.6ms for simple query.

I will try to rerun the baseline with pgbench 15.3 just to make sure the client versions are also the same.

tcpb-like
```bash
postgres@db-yrxscxotxwhggepntnqk:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 52901
number of failed transactions: 0 (0.000%)
latency average = 112.694 ms
initial connection time = 627.669 ms
tps = 887.357526 (without initial connection time)
```
select-only
```bash
postgres@db-yrxscxotxwhggepntnqk:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 899647
number of failed transactions: 0 (0.000%)
latency average = 6.608 ms
initial connection time = 624.868 ms
tps = 15132.534973 (without initial connection time)
```
simple-update
```bash
postgres@db-yrxscxotxwhggepntnqk:/home/ubuntu$ pgbench -c 100 -T 60 -b simple-update
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 333542
number of failed transactions: 0 (0.000%)
latency average = 17.834 ms
initial connection time = 628.916 ms
tps = 5607.339064 (without initial connection time)
```

Comment (sweatybridge) [reply ansible/tasks/setup-postgres.yml:98]: @pcnc @darora is this one good to go? I will resolve the merge conflicts one approved

File: ansible/tasks/setup-supabase-internal.yml
Comment (pcnc) [inline ansible/tasks/setup-supabase-internal.yml:83]: If i understand correctly, then supautils also gets installed through the .deb package, right?

**Review by pcnc — APPROVED** [review]
Just to keep a trail related to performance-improvement work, benchmarks comparing current prod/PPA/current PR build performance can be found [here](https://github.com/supabase/postgres/pull/655):


| AMI Build       | TPCB-like  | select-only | simple-update | Flags        |
|-----------------|------------|-------------|---------------|--------------|
| `15.1.0.75`     | 150.575 ms | 6.591 ms    | 17.112 ms     | -mcpu=native |
| `15.1.0.86-rc1` | 112.694 ms | 6.608 ms    | 17.834 ms     | Official PPA |
| `15.1.0.89-rc1` | 109.967 ms | 6.126 ms    | 16.505 ms     | PPA + mcpu   |

File: ansible/tasks/setup-supabase-internal.yml
Comment (sweatybridge) [reply ansible/tasks/setup-supabase-internal.yml:83]: yup, supautils releases deb package now, similar to pg_graphql.

**Pull request merged**
